### ISSUE TITLE
As a content editor, I want duplicate joined documents to be automatically merged without losing their unique metadata, so that I don't have to merge them manually.

### ISSUE THREAD

**Issue by thatbudakguy** [Princeton-CDH/geniza#170]
# notes for testing
check a few different groups from the spreadsheet to confirm:
- [x] primary document should have all secondary document PGPIDs listed in historic PGPIDs
- [x] historic PGPID should be displayed in admin document edit form
- [x] all tags from any documents in a merged group should be present on the primary document
- [x] all descriptions (unless they are an exact duplicate) should be added to the description with a note about the source document
- [x] all footnotes from all documents in the merged group should be associated with the primary document; any footnotes that are exactly equal should only show up once
- [x] review note should indicate that the document has been merged
- [x] all primary and secondary languages from any documents in the group should be associated with the main document
- [x] all log entries from all documents in the merged group should be re-associated with the primary document (current behavior, need to decide if we want to revisit)
- [x] there should be a new log entry documenting the merge
 
also:
- [x] confirm that no action was taken for rows in the spreadsheet where action is not MERGE
---- 

### Is your feature request related to a problem? Please describe.
Because of the way the current PGP system records joins, there are many duplicate entries for joined documents in the metadata spreadsheet (see https://github.com/Princeton-CDH/geniza/issues/165). Over time, these entries have accumulated unique metadata describing the particular shelfmarks that make up the join.

The new system represents each document with a single PGPID, including joins. We need to merge the duplicate entries, but we want to preserve the unique fragment-level metadata that they have accumulated.

### Describe the solution you'd like
This could be implemented as a management command, and would be easier to accomplish once the data has already been migrated to the database.

In order to support this work, we need to be able to **track historic PGPIDs of a document**, because some PGPIDs will "go away" once their document is merged. We still want to be able to find the relevant information using that PGPID and redirect to the newly merged document.

The logic would be something like:

- [x] Find all documents that are joins, with duplicate shelfmarks, and where the duplicates also have the same document type
- [x] Add the PGPID of each of the duplicates to the historic PGPIDs list of the first/"original" document (use the lowest numeric pgpid)
- [x] Add all the unique tags from each of the duplicates to the tags of the primary document
- [x] Add the descriptions of each of the duplicates (if unique) to the description of the primary document, with a note identifying each description by shelfmark, e.g. `Description for ENA 2262.5: `
- [x] Consolidate edit history — reassociate any unique events for each duplicates with the primary document; delete any duplicates
- [x] Consolidate footnotes — reassociate any unique footnotes with the primary document; delete duplicates
- [x] Add a review note to the first document explaining that it was merged from a set of duplicate joins
- [x] Add old pgpid to admin change view.
- [x] implement `__eq__` method for footnote model
- [x] check if footnotes are equal when migrating documents
- [x] add pgpid to log entry action when moving; use format [PGPID ###]
- [x] unit tests for missing coverage
- [x] when merging footnotes, if footnotes match except for transcription and one has no content, remove the footnote with no transcription content

### Describe alternatives you've considered
The `import_data` command is already rather complicated, so adding this logic there would be tough. There are about 900 documents that need merging, which would make it daunting to do manually.

**Comment by rlskoeser** [comment]
upgrading to 5pts

**Comment by rlskoeser** [comment]
Discussed with @richmanrachel what to do with the log entries when merging — leave them unmerged? label them so it's clear which events belong to which item? — and we want to postpone the decision until we see what it looks like when we run the current merge logic on the test site. (Current logic is to move any non-duplicated log entries over to the primary document.)

**Comment by rlskoeser** [comment]
Ran with a CSV export generated from the current merge report google sheet document. Here is the script output:
```
$ ./manage.py merge_documents merge /tmp/merge-report.csv
Loaded 235 merge rows from /tmp/merge-report.csv
Primary document 31266 not found
Successfully merged 98 groups; skipped 1
```
I looked at the spreadsheet, and 31266 shows twice — by the time the script hits it the second time, it's already been merged in the previous group where it's listed.

**Comment by richmanrachel** [comment]
> historic PGPID should be displayed in admin document edit form
- What do you mean by this? I see it as a static entry but not a changeable one.

> all footnotes from all documents in the merged group should be associated with the primary document; any footnotes that are exactly equal should only show up once
- This isn't working yet, see ID 2493 with double Typed Text footnote

> all primary and secondary languages from any documents in the group should be associated with the main document
- I don't think this is testable as so few docs have associated languages yet.

**Comment by richmanrachel** [comment]
A screenshot of the log history of a 3-part merge to help us determine how log entries should be merged:
![image](https://user-images.githubusercontent.com/68028846/129968147-635c7b79-72d7-4be5-8677-1cc3a0c63a7d.png)
If it's easy, would it be possible to attach PGPIDs to the action section so we can see which one is which?

**Comment by rlskoeser** [comment]
> > historic PGPID should be displayed in admin document edit form
> 
> * What do you mean by this? I see it as a static entry but not a changeable one.

Correct, sorry I wasn't clearer. It should be displayed and read only — old PGPID should not be editable, and should only ever change via merge functions (merge script or planned admin merge feature)

**Comment by rlskoeser** [comment]
> > all footnotes from all documents in the merged group should be associated with the primary document; any footnotes that are exactly equal should only show up once
> 
> * This isn't working yet, see ID 2493 with double Typed Text footnote

If you look at the footnotes in the other view, you'll actually see that they differ because they have _different_ transcription content https://test-geniza.cdh.princeton.edu/admin/footnotes/footnote/?q=2493 

Even though transcription content is provisional right now, we don't want to lose it in the merge. I think it's ok to have duplicate footnotes in those cases for now, what do you think? We could develop some kind of workflow or convention for adding notes so people working with the data will know why there are two footnotes that look the same

**Comment by rlskoeser** [comment]
> If it's easy, would it be possible to attach PGPIDs to the action section so we can see which one is which?

Thanks for the screenshot, super helpful! Yes, I like the idea of adding PGPIDs to the action message. 

How about adding to the end in this format `[PGPID ###]` ?

Is it ok if actions on the primary record do not get a PGPID added?

**Comment by rlskoeser** [comment]
> > all primary and secondary languages from any documents in the group should be associated with the main document
> 
> * I don't think this is testable as so few docs have associated languages yet.

It would be good to test this. When we're ready to do another test run, before I run the script,  you can edit a couple the records on the test site that will be merged and add some languages so we can see what happens.

**Comment by richmanrachel** [comment]
> How about adding to the end in this format [PGPID ###] ? Is it ok if actions on the primary record do not get a PGPID added?
- Yes, I like both of these solutions.

Should I edit languages now or wait till the test site gets updated again to be current?

**Comment by rlskoeser** [comment]
> Should I edit languages now or wait till the test site gets updated again to be current?

let's wait until we're ready to run it with the change for the log entries, in case we want to refresh the database again for some reason

**Comment by richmanrachel** [comment]
I changed the current database metadata for a few of the docs in the merge lineup to ensure that languages transfer properly for the next round of testing:
- Added Hebrew to PGPID 460 as the non-primary doc.
- Added Hebrew to 533 as primary and 9029 as non-primary to check for redundancies. 
[These languages are correct and not just a test]

**Comment by rlskoeser** [comment]
Updated test site with a current copy of production data, confirmed that the languages were present on the PGPIDs you noted (although I didn't see a secondary language), and then ran the merge script again (using the same CSV I tested with before):
```
$ ./manage.py merge_documents merge /tmp/merge-report.csv
Loaded 235 merge rows from /tmp/merge-report.csv
Primary document 31266 not found
Successfully merged 98 groups; skipped 1
```

Please test to check how languages are being merged and that merged log entries are labeled with original PGPID

**Comment by richmanrachel** [comment]
@rlskoeser - It seems like footnotes are doubling. Here's an example from ID 533:
![image](https://user-images.githubusercontent.com/68028846/130658602-f04f8dc1-ec94-41e4-beeb-c017612de630.png)

**Comment by richmanrachel** [comment]
Not exactly sure why my 1:19pm entry has a bunch of brackets, but I like the inclusion of old PGPID here! 
![image](https://user-images.githubusercontent.com/68028846/130658832-386be9e0-c62a-4902-8fe1-4030277fe8db.png)

**Comment by richmanrachel** [comment]
@rlskoeser, I confirmed on ID 9804 that both footnote entries were doubled. So in sum, languages work, but footnotes still need work.

**Comment by rlskoeser** [comment]
@richmanrachel ran the latest version of the script with the data from **Merging Round 2**; here's the output from the command line:
```
$ ./manage.py merge_documents merge /tmp/Merging\ Round\ 2\ -\ merge-report.csv
Loaded 82 merge rows from /tmp/Merging Round 2 - merge-report.csv
Successfully merged 41 groups; skipped 0
```

Please review the merged documents & check that footnotes are being merged the way we want them now.

**Comment by richmanrachel** [comment]
@rlskoeser - Adding footnotes from different sources together is working (for eg. see 492 -> 5583). And I haven't run into any doubled ones yet! 

Should we run the original CSV again with the joins so I can check the ones I flagged earlier?

**Comment by richmanrachel** [comment]
(This spreadsheet: https://docs.google.com/spreadsheets/d/1cN3XK224O0rZhlijuN57Rz04Xr-bcMlDGnBD3cAqcAE/edit#gid=1575945226 )

**Comment by rlskoeser** [comment]
@richmanrachel thanks for the link! Downloaded as csv and ran the script, here's the output:
```
$ ./manage.py merge_documents merge /tmp/merge-report\ -\ merge-report.csv
Loaded 235 merge rows from /tmp/merge-report - merge-report.csv
Primary document 31266 not found
Successfully merged 98 groups; skipped 1
```

**Comment by richmanrachel** [comment]
@rlskoeser - No doubling on 9804, hooray! Do you know which group skipped?

**Comment by rlskoeser** [comment]
@richmanrachel pretty sure it's the primary document not found — 31266

**Comment by richmanrachel** [comment]
@rlskoeser - Everything looks good... even 31266 seems to have worked?
![image](https://user-images.githubusercontent.com/68028846/134069769-d74e46ba-56c7-4fd7-bb29-ba2ff9f113d7.png)

**Comment by richmanrachel** [comment]
@rlskoeser Should I close or should I be concerned about the fact that the script says it wasn't found?

**Comment by rlskoeser** [comment]
@richmanrachel I think the not found error is a data problem, not a code issue. If all the merged documents look good then I think we're ok to close this finally
